### PR TITLE
8310948: Fix ignored-qualifiers warning in Hotspot

### PR DESCRIPTION
**Pull request by djelinski**
Please review this attempt to fix ignored-qualifiers warning.

Example warnings:
```
src/hotspot/share/oops/method.hpp:413:19: warning: 'volatile' type qualifier on return type has no effect [-Wignored-qualifiers]
   CompiledMethod* volatile code() const;
                   ^~~~~~~~~
```
```
src/hotspot/share/jfr/periodic/jfrModuleEvent.cpp:65:20: warning: type qualifiers ignored on cast result type [-Wignored-qualifiers]
    65 |   event.set_source((const ModuleEntry* const)from_module);
       |                    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

The proposed fix removes the ignored qualifiers.
In a few AD files I replaced `const` with `constexpr` where I noticed that the method is returning a compile-time constant, and other platforms use `constexpr` on the same method.

Release, debug and slowdebug builds on Aarch64 / x64 and Mac / Linux complete without errors. Cross-compile GHA builds also pass.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8310948](https://bugs.openjdk.org/browse/JDK-8310948): Fix ignored-qualifiers warning in Hotspot (**Enhancement** - P4)

### Reviewers
 * [Kim Barrett](https://openjdk.org/census#kbarrett) (@kimbarrett - **Reviewer**)
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/14674/head:pull/14674` \
`$ git checkout pull/14674`

Update a local copy of the PR: \
`$ git checkout pull/14674` \
`$ git pull https://git.openjdk.org/jdk.git pull/14674/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 14674`

View PR using the GUI difftool: \
`$ git pr show -t 14674`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/14674.diff">https://git.openjdk.org/jdk/pull/14674.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/14674#issuecomment-1609675270)

### CODE REVIEW

**Comment by dholmes-ora** [comment]
> warning: 'volatile' type qualifier on return type has no effect

Can't say that I understand that warning. If I pass in a volatile pointer and return the same pointer then I would not expect to lose the volatile aspect of it. @kbarrett can you explain this one to me?

**Review by kimbarrett — APPROVED** [review]
Looks good.  Though mind-numbingly large; breaking something like this up into easier
to digest chunks can make reviewing easier.

**Comment by djelinski** [comment]
David: I think this part of the spec is relevant here:
> A non-class non-array prvalue cannot be [cv-qualified](https://en.cppreference.com/w/cpp/language/cv), [...]. (Note: a function call or cast expression may result in a prvalue of non-class cv-qualified type, but the cv-qualifier is generally immediately stripped out.)

[source](https://en.cppreference.com/w/cpp/language/value_category)
given that the cv qualifiers are immediately stripped by the compiler, there's no point in providing them.

In the particular volatile pointer case: the function performs a volatile read to get the pointer value (address). That address can then be used in a non-volatile manner.

Kim: I realize that it's a big change, so thank you very much for reviewing it anyway! I was prepared to split it up, just wanted to know if this warning is something we want to fix.

**Comment by dean-long** [comment]
In the example, was the intention perhaps `volatile CompiledMethod*` instead of `CompiledMethod* volatile`?

**Comment by djelinski** [comment]
I don't think that was the intention. There's a comment on `Method::_code` stating that the pointer can change at any point (so the pointer was supposed to be volatile), and `class CompiledMethod` does not have any `volatile` - qualified methods.

There are similar cases where `const` pointer to a class was returned involving `JavaThread`, `ReferenceProcessor`, `PSCardTable` and `ShenandoahHeapRegion`; I suppose we could review if `const` could be added to the returned class, but that's outside of the scope of this PR.

**Comment by djelinski** [comment]
Can I get another +1 on this? or should I proceed with splitting?

File: src/hotspot/cpu/aarch64/aarch64.ad [hunk 1]
```diff
@@ -2285,7 +2285,7 @@ int HandlerImpl::emit_deopt_handler(CodeBuffer& cbuf)
 
 //=============================================================================
 
-const bool Matcher::match_rule_supported(int opcode) {
```
Comment (dholmes-ora) [inline src/hotspot/cpu/aarch64/aarch64.ad:2288]: Have to wonder if these were all meant to be `bool Match:xxx() const {`?

**Review by dholmes-ora — APPROVED** [review]
I will approve this as-is but have to wonder whether many of these cases of const return types were intending to declare const functions?

**Comment by dholmes-ora** [comment]
P.S. Forgot to say thanks for dealing with this!

File: src/hotspot/cpu/aarch64/aarch64.ad (hunk 1 above)
Comment (djelinski) [reply src/hotspot/cpu/aarch64/aarch64.ad:2288]: Yes, I think that may have been the original intent. I'll add const on these functions.

Comment (dholmes-ora) [reply src/hotspot/cpu/aarch64/aarch64.ad:2288]: Ah okay :)

**Comment by djelinski** [comment]
Thanks for the reviews!

/integrate

**Pull request closed without merging**